### ISSUE TITLE
Infinite scanning with TLS-Scanner

### ISSUE THREAD

**Issue by manmolecular** [RUB-NDS/TLS-Scanner#59]
Hello.

Sometimes scanning process of TLS-Scanner does not finish - on some particular hosts scanner can work for hours (for example, the longest run until successful results were gained was about 3 hours) or even can get stuck at some point in an infinite loop (especially when TLS-Scanner on the latest steps of scanning, for example, about 21/24 or 22/24 step in console output). 

As I noticed, this behavior does not depend on "-parallelProbes" or "-overallThreads" flags and can be reproduced with the most basic options.

Examples of this behavior can be seen on the next hosts:
```
67.59.100.154:443
185.42.19.133:443
```
And can be reproduced with the next commands:
```
java -jar TLS-Scanner.jar -connect 67.59.100.154:443
java -jar TLS-Scanner.jar -connect 185.42.19.133:443
```

**Comment by ic0ns** [comment]
Hey,
i could scan both hosts - it just took a long time. This is because the padding oracle & bleichenbacher scans failed and the scanner then tries to rescan with longer timeouts in between and the server supports a lot of cipher suites which need testing, but it did eventually terminate for me (Scanned in: 704s, Scanned in: 21s).

**Comment by manmolecular** [comment]
Thanks for your answer, I was able to successfully reproduce scan with `-overallThreads 100` and `-parallelProbes 10` and it was much faster (in my case these hosts were scanned in 516s and 262s, so that is ok). Previously, I used lower values for these flags.

In this case, could you please tell, what values for `overallThreads` and `parallelProbes` are optimal for correct scanning results In the most general case? 

And one more question - for this scans, for example, padding oracle response map looks like this:
```
PaddingOracle Responsemap

TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA       - TLS12     # Error during Scan
TLS_DHE_RSA_WITH_AES_256_CBC_SHA         - TLS12     # Error during Scan
TLS_DHE_RSA_WITH_CAMELLIA_256_CBC_SHA    - TLS12     # Error during Scan
TLS_RSA_WITH_AES_256_CBC_SHA             - TLS12     # Error during Scan
TLS_RSA_WITH_CAMELLIA_256_CBC_SHA        - TLS12     # Error during Scan
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA       - TLS12     # Error during Scan
TLS_DHE_RSA_WITH_AES_128_CBC_SHA         - TLS12     # Error during Scan
TLS_DHE_RSA_WITH_CAMELLIA_128_CBC_SHA    - TLS12     # Error during Scan
TLS_RSA_WITH_AES_128_CBC_SHA             - TLS12     # Error during Scan
TLS_RSA_WITH_CAMELLIA_128_CBC_SHA        - TLS12     # Error during Scan
TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384    - TLS12     # Error during Scan
TLS_DHE_RSA_WITH_AES_256_CBC_SHA256      - TLS12     # Error during Scan
TLS_RSA_WITH_AES_256_CBC_SHA256          - TLS12     # Error during Scan
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256    - TLS12     # Error during Scan
TLS_DHE_RSA_WITH_AES_128_CBC_SHA256      - TLS12     # Error during Scan
TLS_RSA_WITH_AES_128_CBC_SHA256          - TLS12     # Error during Scan
```
Does it depend somehow on these parameters (parallelProbes, overallThreads)? Or these errors connected with something else?

Additional information: 
1. TLS-Attacker and TLS-Scanner successfully passed all tests. 
2. Versions are the latest ones from the master branch of TLS-Attacker and TLS-Scanner repositories (TLS-Attacker 3.2). 
3. Java: openjdk version "1.8.0_222"
4. System: Ubuntu 18.04.2 LTS

**Comment by ic0ns** [comment]
TLDR: The error occurs when the TLS-Scanner _thinks_ that something went wrong during the Handshake. 
Long version:
The test works internally with WorkflowTraces. Meaning that for each Handshake the scanner has to do, it has internally an idea of how the other party will respond to the messages it sends. If after the execution of the handshake the responses look differently, the handshake is marked as "noExecutedAsPlanned" and the execution is repeated ( in the hope that this was just some anomaly and that the next handshake will hopefully work). Some implementations have functional bugs such that you literally can not finish the handshake with the selected parameter version. Other implementations just do something which is perfectly fine and compliant with the RFC, which was just not anticipated by the scanner. In this case, I believe its a CertificateRequest message sent by the server. Therefore the test is marked as failed. This is basically #58 this issue.

**Comment by manmolecular** [comment]
Okay, thanks for the information, it was helpful.

Sorry to interrupt you for the last time, but could you please tell what optimal values would be okay for `overallThreads` and `parallelProbes` flags in general to get correct results faster? 
`-overallThreads 100 -parallelProbes 10`, as you mention?

**Comment by ic0ns** [comment]
It depends on multiple things like the server and the config of said server or the hardware you are running the scanner on. For example servers with a lot of CBC supported ciphersuites generally require more handshakes to be executed, which is speed up with the -overallThreads. I think you just have to experiment what works best for you. If you run with -overallThreads 100 and -parallelProbes 10 with a single Scanner instance you will probably get very fast results, but I can also hear my laptop's fan start when I do so (its not very resource efficient). Some open-source example servers do not support multiple handshakes at the same time, so you gain nothing by adding more threads to the scanner. 
For a single server I don't think that more than 100 overall Threads currently brings any advantage
nor does more than 13 parallel probes. Note that these numbers can change with time as we update the scanner.
If you are doing a lot of bulk testing you can highly speed up your scanning experience by using bigger threadpools and sharing them among the scanners (there is a constructor for that in the TlsScanner.java class).

**Comment by manmolecular** [comment]
Thanks for all your answers.
